### ISSUE TITLE
[Question] How to bind one class to many different dependencies?

### ISSUE THREAD

**Issue by gCardinal** [inversify/InversifyJS#428]
I'll illustrate with an example. Say I have a custom `OAuthClient` for my app that uses a variety of `OAuthProvider` to determine certain information to authenticate and validate users. I would have a basic `OAuthClient` with this as a constructor:

```typescript
import {ValidationHandler} from 'majisti/authentication/ValidationHandler'
import {OAuthProvider} from 'majisti/authentication/OAuthProvider'
import {WindowFacade} from 'utils/WindowFacade'

export default class OAuthClient {
    private provider: OAuthProvider
    private windowFacade: WindowFacade
    private validationHandler: ValidationHandler

    public constructor(provider: OAuthProvider, windowFacade: WindowFacade, validationHandler: ValidationHandler) {
        this.provider = provider
        this.windowFacade = windowFacade
        this.validationHandler = validationHandler
    }
}
```

Now, the part I don't understand how to do with InversifyJS is having multiple different providers for different clients, but without having to duplicate my code. What I mean is that, assuming I inject like this for a client the container would bind to `Types.LinkedInOAuthClient`:

```typescript
import {injectable, inject} from "inversify"

import {ValidationHandler} from 'majisti/authentication/ValidationHandler'
import {OAuthProvider} from 'majisti/authentication/OAuthProvider'
import {WindowFacade} from 'utils/WindowFacade'

@injectable()
export default class OAuthClient {
    private provider: OAuthProvider
    private windowFacade: WindowFacade
    private validationHandler: ValidationHandler

    public constructor(@inject(Types.LinkedInAuthProvider) provider: OAuthProvider,
		@inject(Types.WindowFacade) windowFacade: WindowFacade,
		@inject(Types.JwtValidationHandler) validationHandler: ValidationHandler) {
        this.provider = provider
        this.windowFacade = windowFacade
        this.validationHandler = validationHandler
    }
}
```

Then I'd have to create a new class, just to be able to inject a different provider and, since all providers are different, I can't just group them under a single `Types.OAuthProvider`.

I guess my question is: How can I have my container create, say, 3 different auth clients (one for Google, Facebook and LinkedIn) and inject a different `OAuthProvider` for each?

**Comment by remojansen** [comment]
You need some kind of additional metadata and use a [contextual constraint](https://github.com/inversify/InversifyJS/blob/master/wiki/contextual_bindings.md) based on that metadata. We support some features for this kind of use case like [Tagged Binding](https://github.com/inversify/InversifyJS/blob/master/wiki/tagged_bindings.md) and [Named Binding](https://github.com/inversify/InversifyJS/blob/master/wiki/named_bindings.md). 

In your case you are also going to need a [Factory Injection](https://github.com/inversify/InversifyJS/blob/master/wiki/factory_injection.md). Let's take a look to an example!

You only need one `OAuthClient` implementation:
```ts
import {injectable, inject} from "inversify"
import {ValidationHandler} from 'majisti/authentication/ValidationHandler'
import {OAuthProvider} from 'majisti/authentication/OAuthProvider'
import {WindowFacade} from 'utils/WindowFacade'

@injectable()
export default class OAuthClient {

    private provider: OAuthProvider
    private windowFacade: WindowFacade
    private validationHandler: ValidationHandler

    public constructor(
        @inject(Types.OAuthProvider) provider: OAuthProvider,
        @inject(Types.WindowFacade) windowFacade: WindowFacade,
        @inject(Types.JwtValidationHandler) validationHandler: ValidationHandler
    ) {
        this.provider = provider
        this.windowFacade = windowFacade
        this.validationHandler = validationHandler
    }
}
```

I don't know what kind of app you are creating but I will use a REST API with [inversify-express-utils](https://github.com/inversify/inversify-express-utils) as an example. You will need to inject a `Factory<OAuthClient>` into a Controller:

```ts
@injectable()
@Controller('/auth')
class LoginController() {

    private _oAuthClientFactory: Factory<OAuthClient>;

    public constructor(@inject("Factory<OAuthClient>") oAuthClientFactory: Factory<OAuthClient>) {
        this._oAuthClientFactory = oAuthClientFactory;
    }

    @Post('/login')
    public login(request: Request) {
        let name = request.header.provider; // read provider from a header
        let oAuthClient = this.oAuthClientFactory(name); // get client using provider as metadata
    }
}
```

Notice how the factory uses a header as an argument. This header is used as metadata. 

> Controller -> Client (Parent) -> Provider

The metadata is used in some contectual constraints. Notice the `whenParentNamed` constraints and how it is used in the `Factory<OAuthClient>`:

```ts
let container = new Container();
container.bind<OAuthClient>("OAuthClient").to(OAuthClient);

container.bind<OAuthProvider>("OAuthProvider").to(LinkedInAuthProvider).whenParentNamed("LinkedIn");
container.bind<OAuthProvider>("OAuthProvider").to(FacebookAuthProvider).whenParentNamed("Facebook");
container.bind<OAuthProvider>("OAuthProvider").to(GoogleAuthProvider).whenParentNamed("Google");

container.bind<interfaces.Factory<OAuthClient>>("Factory<OAuthClient>").toFactory<OAuthClient>((context: interfaces.Context) => {
    return (named: string) => {
        return context.container.getNamed<OAuthClient>("OAuthClient", named);
    };
});
```

I hope this helps.

Please let e know if I can close this issue.

**Comment by gCardinal** [comment]
Thanks a lot for the answer, it helped a lot. I'm on 2.0.1 so I had to figure out a few changes, but overall your explanation was very clear!

I'm having one more problem though.

Following your example, I'm getting a `Uncaught Error: No bindings found for serviceIdentifier: Symbol(OAuthClient)`. I'll paste my code here and then explain what I tried to resolve it.

```typescript
// The OAuthClient, which is pretty much exactly the same as in your example
import {injectable, inject} from 'inversify'
import {ValidationHandler} from 'majisti/authentication/ValidationHandler'
import Types from 'di/Types'
import WindowFacade from 'utils/WindowFacade'
import OAuthProvider from 'majisti/authentication/OAuthProvider'

@injectable()
export default class OAuthClient {
    private provider: OAuthProvider
    private windowFacade: WindowFacade
    private validationHandler: ValidationHandler

    public constructor(@inject(Types.OAuthProvider) provider: OAuthProvider,
                       @inject(Types.WindowFacade) windowFacade: WindowFacade,
                       @inject(Types.JwtValidationHandler) validationHandler: ValidationHandler) {
        this.provider = provider
        this.windowFacade = windowFacade
        this.validationHandler = validationHandler
    }
}
```
And then, in the kernel module, I set it up like this:

```typescript
import {KernelModule} from 'inversify'
import interfaces from 'inversify/dts/interfaces/interfaces'
import OAuthClient from 'majisti/authentication/OAuthClient'
import Types from 'di/Types'
import Configuration from 'Configuration'
import OAuthProvider from 'majisti/authentication/OAuthProvider'
import Routes from 'routing/Routes'
import Factory = interfaces.Factory
import Context = interfaces.Context

export default new KernelModule((bind) => {
    const urls = Configuration.urls

    const linkedInProvider = new OAuthProvider(
        urls.linkedIn.baseUri() + urls.linkedIn.authorizationEndpoint(),
        urls.api.baseUri() + urls.api.validationUri(),
        urls.linkedIn.clientId(),
        urls.linkedIn.callbackUrl(Routes.CONVERSATIONS),
        'r_basicprofile r_emailaddress',
        '83740234234'
    )

    bind<OAuthClient>(Types.OAuthClient).to(OAuthClient)

    bind<OAuthProvider>(Types.OAuthProvider)
        .toConstantValue(linkedInProvider)
        .whenParentNamed(Types.LinkedInOAuthClient)

    bind<Factory<OAuthClient>>(Types.OAuthClientFactory)
        .toFactory<OAuthClient>(
            (context: Context) => {
                return (name: string) => {
                    return context.kernel.getNamed<OAuthClient>(Types.OAuthClient, name)
                }
            })
})
```

This isn't working, I get the above error. I thought it might be because my `OAuthClient` service was not named, so I tried doing

```typescript
bind<OAuthClient>(Types.OAuthClient).to(OAuthClient).whenTargetNamed(Types.GoogleOAuthClient)
bind<OAuthClient>(Types.OAuthClient).to(OAuthClient).whenTargetNamed(Types.LinkedInOAuthClient)

// Instead of
bind<OAuthClient>(Types.OAuthClient).to(OAuthClient)
```

But now I'm wondering if this isn't problematic since there's two implementations of the same class? Any idea what I'm doing wrong?

**Comment by remojansen** [comment]
Hi @gCardinal I saw you posted a comment but I can't see it now? 

<img width="677" alt="screen shot 2016-12-01 at 19 01 08" src="https://cloud.githubusercontent.com/assets/10656223/20807943/b30ef8b6-b7f8-11e6-9920-277a207026ba.png">

**Comment by gCardinal** [comment]
I was hoping I had been fast enough that you didn't get the notification!

Yes I did post something, I was having a problem I couldn't figure out and thought maybe my implementation of your solution was flawed, but it turns out it was simply a silly mistake that I eventually fixed.

That being said, I got the factory working! Your example was great and helped me understand InversifyJS a bit more. It's still a bit weird to me how some things are handled, I'm just too used to the way Symfony does dependency injection, but overall this project is growing on me! Thank you for the help and the great library.
